api/types/container: refactor to use strings.Cut, DRY, move tests and fix validation

thaJeztah opened the pull request:
### hostconfig: move unit tests to api/types/container

These types were moved to api/types/container in https://github.com/moby/moby/commit/7ac4232e70fe7cf7318333cd0890db7f95663079 (https://github.com/moby/moby/pull/18762), but the unit-tests for them were not moved. This patch moves the unit-tests back together with the types.

### api/types/container: rewrite tests to use subtests and asserts

### api/types/container: use strings.Cut() and DRY

### api/types/container: fix validation for UTSMode, UsernsMode, PidMode

The IPCMode type was added in 497fc8876ede9924f61c0eee4dfadd71e5d9f537 (https://github.com/moby/moby/pull/9074), and from that patch, the intent was to allow `host` (without `:`), `""` (empty, default) or `container:<container ID>`, but the `Valid()` function seems to be too relaxed and accepting both `:`, as well as `host:<anything>`. No unit-tests were added in that patch, and integration-tests only tested for valid values.

Later on, `PidMode`, and `UTSMode` were added in 23feaaa240853c0e7f9817f8c2d272dd1c93ac3f ( https://github.com/moby/moby/pull/9339 / https://github.com/moby/moby/pull/10080) and f2e5207fc989288ad136d48222df8e7754eb0e9b (https://github.com/moby/moby/pull/12667), both of which were implemented as a straight copy of the `IPCMode` implementation, copying the same bug.

Finally, commit d4aec5f0a680b6b01bb720830451a93c6ec398e6 (https://github.com/moby/moby/pull/14134) implemented unit-tests for these types, but testing for the wrong behavior of the implementation.

This patch updates the validation to correctly invalidate `host[:<anything>]` and empty (`:`) types.

thaJeztah committed:
api/types/container: use strings.Cut() and DRY

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
hostconfig: move unit tests to api/types/container

These types were moved to api/types/container in 7ac4232e70fe7cf7318333cd0890db7f95663079,
but the unit-tests for them were not moved. This patch moves the unit-tests back together
with the types.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
api/types/container: rewrite tests to use subtests and asserts

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
api/types/container: fix validation for UTSMode, UsernsMode, PidMode

The IPCMode type was added in 497fc8876ede9924f61c0eee4dfadd71e5d9f537, and from
that patch, the intent was to allow `host` (without `:`), `""` (empty, default)
or `container:<container ID>`, but the `Valid()` function seems to be too relaxed
and accepting both `:`, as well as `host:<anything>`. No unit-tests were added
in that patch, and integration-tests only tested for valid values.

Later on, `PidMode`, and `UTSMode` were added in 23feaaa240853c0e7f9817f8c2d272dd1c93ac3f
and f2e5207fc989288ad136d48222df8e7754eb0e9b, both of which were implemented as
a straight copy of the `IPCMode` implementation, copying the same bug.

Finally, commit d4aec5f0a680b6b01bb720830451a93c6ec398e6 implemented unit-tests
for these types, but testing for the wrong behavior of the implementation.

This patch updates the validation to correctly invalidate `host[:<anything>]`
and empty (`:`) types.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
@rumpl @corhere PTAL

corhere reviewed (approved):
Per [Hyrum's Law](https://www.hyrumslaw.com/) there is inevitably going to be someone (likely with a buggy script) who is going to be upset that we "broke" them. We should call this fix out in the changelog since it is technically a breaking API change.

thaJeztah commented on api/types/container/hostconfig.go:
```diff
@@ -463,3 +432,13 @@ type HostConfig struct {
+// a "false" if the value does not have a "container:" prefix. Further validation
+// of the returned, including checking if the value is empty, should be handled
+// by the caller.
+func containerID(val string) (idOrName string, ok bool) {
+	k, idOrName, ok := strings.Cut(val, ":")
+	return idOrName, ok && k == "container"
+}
```
Didn't want to go all overboard on DRY 😂 I think I'll keep that for a future exercise;

corhere replied:
Fair. I only really looked into this to sate my curiosity.

thaJeztah commented on api/types/container/hostconfig.go:
```diff
@@ -183,36 +180,27 @@ func (n UsernsMode) IsPrivate() bool {
-		return false
-	}
-	return true
+	return string(n) == "" || n.IsHost()
```
LOL, I was actually looking at those after I rebased and pushed ("was this actually needed?"); I'll have a look.

thaJeztah commented:
> Per [Hyrum's Law](https://www.hyrumslaw.com/) there is inevitably going to be someone (likely with a buggy script) who is going to be upset that we "broke" them. We should call this fix out in the changelog since it is technically a breaking API change.

@corhere I added `impact/changelog`, and removed the redundant `string()` casts; PTAL 🤗
